aten/src/ATen/native/SharedReduceOps.h

```diff
@@ -122,6 +122,31 @@ struct WelfordOps {
   }
 };
 
+template <typename acc_t>
+struct NanProdOps {
+  inline C10_DEVICE acc_t reduce(acc_t a, acc_t b, int64_t /*idx*/) const {
+    return combine(a, b);
+  }
+
+  inline C10_DEVICE acc_t combine(acc_t a, acc_t b) const {
+    return (at::_isnan(a) ? acc_t{1} : a) * (at::_isnan(b) ? acc_t{1} : b);
```

Comment:
Also: `np.nanprod(float('nan'))` gives `1.0` as the output. Is that a desired output? I'm not sure if we want to return 1 if it's just `nan` there. I'll let the upstream decide on this.

Reply:
That makes sense since nanprod is conceptually equivalent to `nan_to_num(t, nan=1)` followed by `prod(t)`.